cpp/include/cuvs/preprocessing/quantization.hpp

```diff
@@ -0,0 +1,222 @@
+ * @tparam QuantI quantized type of data after transform
+ *
+ */
+template <typename T, typename QuantI>
+struct ScalarQuantizer {
+  T min_;
+  T max_;
+};
+
+/**
+ * @brief Initializes a scalar quantizer to be used later for quantizing the dataset.
+ *
+ * Usage example:
+ * @code{.cpp}
+ * raft::handle_t handle;
+ * cuvs::preprocessing::quantization::sq_params params;
+ * auto quantizer = cuvs::preprocessing::quantization::train_scalar<float, int8_t>(handle, params,
+ * dataset);
+ * @endcode
+ *
+ * @tparam T data element type
+ * @tparam QuantI quantized type of data after transform
+ *
+ * @param[in] res raft resource
+ * @param[in] params configure scalar quantizer, e.g. quantile
+ * @param[in] dataset a row-major matrix view on device
+ *
+ * @return ScalarQuantizer
+ */
+template <typename T, typename QuantI>
+ScalarQuantizer<T, QuantI> train_scalar(raft::resources const& res,
+                                        const sq_params params,
+                                        raft::device_matrix_view<const T, int64_t> dataset);
+
+/**
+ * @brief Initializes a scalar quantizer to be used later for quantizing the dataset.
+ *
+ * Usage example:
+ * @code{.cpp}
+ * raft::handle_t handle;
+ * cuvs::preprocessing::quantization::sq_params params;
+ * auto quantizer = cuvs::preprocessing::quantization::train_scalar<float, int8_t>(handle, params,
+ * dataset);
+ * @endcode
+ *
+ * @tparam T data element type
+ * @tparam QuantI quantized type of data after transform
+ *
+ * @param[in] res raft resource
+ * @param[in] params configure scalar quantizer, e.g. quantile
+ * @param[in] dataset a row-major matrix view on host
+ *
+ * @return ScalarQuantizer
+ */
+template <typename T, typename QuantI>
+ScalarQuantizer<T, QuantI> train_scalar(raft::resources const& res,
+                                        const sq_params params,
+                                        raft::host_matrix_view<const T, int64_t> dataset);
+
+/**
+ * @brief Applies quantization transform to given dataset
+ *
+ * Usage example:
+ * @code{.cpp}
+ * raft::handle_t handle;
+ * cuvs::preprocessing::quantization::sq_params params;
+ * auto quantizer = cuvs::preprocessing::quantization::train_scalar<float, int8_t>(handle, params,
+ * dataset); auto quantized_dataset = raft::make_device_matrix<int8_t, int64_t>(handle, samples,
+ * features); cuvs::preprocessing::quantization::transform(handle, quantizer, dataset,
+ * quantized_dataset.view());
+ * @endcode
+ *
+ * @tparam T data element type
+ * @tparam QuantI quantized type of data after transform
+ *
+ * @param[in] res raft resource
+ * @param[in] quantizer a scalar quantizer
+ * @param[in] dataset a row-major matrix view on device
+ * @param[out] out a row-major matrix view on device
+ *
+ */
+template <typename T, typename QuantI>
+void transform(raft::resources const& res,
+               const ScalarQuantizer<T, QuantI>& quantizer,
+               const raft::device_matrix_view<const T, int64_t> dataset,
```

Comment:
We don't usually use const device_matrix_view because it is already a simple immutable view type. Instead please just use device_matrix_view<const T...>